planner/cascades/optimize.go

```diff
@@ -48,6 +51,111 @@ func onPhaseExploration(sctx sessionctx.Context, g *Group) error {
 	return errors.New("the onPhaseExploration() of the cascades planner is not implemented")
 }
 
+// onPhaseImplementation starts implementating physical operators from given root Group.
 func onPhaseImplementation(sctx sessionctx.Context, g *Group) (plannercore.Plan, error) {
-	return nil, errors.New("the onPhaseImplementation() of the cascades planner is not implemented")
+	prop := &property.PhysicalProperty{
+		ExpectedCnt: math.MaxFloat64,
+	}
+	// TODO replace MaxFloat64 costLimit by variable from sctx, or other sources.
+	impl, err := implGroup(g, prop, math.MaxFloat64)
+	if err != nil {
+		return nil, err
+	}
+	if impl == nil {
+		return nil, plannercore.ErrInternal.GenWithStackByArgs("Can't find a proper physical plan for this query")
+	}
+	return impl.getPlan(), nil
+}
+
+// implGroup picks one implementation with lowest cost for a Group, which satisfies specified physical property.
+func implGroup(g *Group, reqPhysProp *property.PhysicalProperty, costLimit float64) (Implementation, error) {
+	groupImpl := g.getImpl(reqPhysProp)
+	if groupImpl != nil {
+		if groupImpl.getCost() <= costLimit {
+			return groupImpl, nil
+		}
+		return nil, nil
+	}
+	// Handle implementation rules for each equivalent GroupExpr.
+	var cumCost float64
+	var childCosts []float64
+	outCount := math.Min(g.prop.Stats.RowCount, reqPhysProp.ExpectedCnt)
+	for elem := g.equivalents.Front(); elem != nil; elem = elem.Next() {
+		curExpr := elem.Value.(*GroupExpr)
+		impls, err := implGroupExpr(curExpr, reqPhysProp)
+		if err != nil {
+			return nil, err
+		}
+		for _, impl := range impls {
+			cumCost = 0.0
+			childCosts = childCosts[:0]
+			for i, childGroup := range curExpr.children {
+				childImpl, err := implGroup(childGroup, impl.getPlan().GetChildReqProps(i), costLimit-cumCost)
+				if err != nil {
+					return nil, err
+				}
+				if childImpl == nil {
+					impl.setCost(math.MaxFloat64)
+					break
+				}
+				childCost := childImpl.getCost()
+				childCosts = append(childCosts, childCost)
+				cumCost = cumCost + childCost
+				impl.getPlan().Children()[i] = childImpl.getPlan()
+			}
+			if impl.getCost() == math.MaxFloat64 {
+				continue
+			}
+			cumCost = impl.calcCost(outCount, childCosts, curExpr.children...)
+			if cumCost > costLimit {
+				continue
+			}
+			if groupImpl == nil || groupImpl.getCost() > cumCost {
+				groupImpl = impl
+			}
+		}
+	}
+	// Handle enforcing rules for required physical property.
+	for _, rule := range GetEnforcerRules(reqPhysProp) {
+		newProps := rule.NewProperties(reqPhysProp)
+		for _, newReqPhysProp := range newProps {
+			childImpl, err := implGroup(g, newReqPhysProp, costLimit)
```

Comment:
For enforcers, the cost calculation donot need to know the cost of child implementations, only the logical property of child group is needed. We can firstly calculate the self cost of the enforcer itself and cut `costLimit`, then search the best implementations of child groups.

Reply:
Got it, I considered `RootEnforcer` here before, since it is removed, we can change it to generate sort first. The comments and signature of `OnEnforce` need change as well, I will update it.